weave/trace_server/execution_runner/README.md

```diff
@@ -0,0 +1,284 @@
+## Architecture
+
+The system uses a three-layer architecture:
+
+1. **Main Process** (ClickHouse Trace Server)
+   - Receives incoming requests
+   - Delegates to RunAsUser for isolated execution
+   - Maintains the actual trace server instance
+
+2. **Worker Thread** (in Main Process)
+   - Created by `generate_child_process_trace_server_args()`
+   - Holds the wrapped trace server instance
+   - Processes requests from child via queues
+   - Returns responses back through queues
+
+3. **Child Process** (Isolated User Context)
+   - Spawned by RunAsUser for each execution
+   - Has no direct access to parent's memory
+   - Communicates only through multiprocessing queues
+   - Runs with a user-scoped WeaveClient
+
+## Execution Architecture
+
+### Layer Architecture (User Script Perspective)
+
+This diagram shows how the user script (`run_model`) operates within multiple layers of abstraction:
+
+```mermaid
+graph TB
+    subgraph "Child Process Environment"
+        subgraph "User Script Context"
+            US[User Script<br/>run_model.py]
+            
+            subgraph "Layer 3: Client Layer"
+                USC[user_scoped_client]
+                USC_DESC["• Special project_id: __SERVER__/actual_project<br/>• Scoped to user context<br/>• Provides WeaveClient interface"]
+            end
+            
+            subgraph "Layer 2: Server Proxy Layer"
+                CTS[CrossProcessTraceServer]
+                CTS_DESC["• Proxy to parent process<br/>• Communicates via queues<br/>• Implements TraceServerInterface"]
+            end
+            
+            subgraph "Layer 1: Request Processing"
+                REQ[Externalized Request]
+                REQ_DESC["• Internal refs → External refs<br/>• weave:///entity/project → weave:///__SERVER__/project<br/>• Prevents cross-project access"]
+            end
+        end
+    end
+    
+    subgraph "Parent Process Environment"
+        subgraph "Worker Thread"
+            ETS[UserInjectingExternalTraceServer]
+            ETS_DESC["• Wraps ClickHouse server<br/>• Injects wb_user_id<br/>• Validates project_id"]
+            
+            CH[ClickHouse<br/>Trace Server]
+            CH_DESC["• Original server<br/>• Database access<br/>• Core functionality"]
+        end
+    end
+    
+    %% Connections
+    US -->|uses| USC
+    USC -->|backed by| CTS
+    CTS -.->|queues| ETS
+    ETS -->|wraps| CH
+    
+    %% Request flow
+    REQ -->|provided to| US
+    
+    %% Styling
+    style US fill:#e8f5e9
+    style USC fill:#f3e5f5
+    style CTS fill:#fff3e0
+    style ETS fill:#fce4ec
+    style CH fill:#e3f2fd
+    
+    classDef descBox fill:#f5f5f5,stroke:#999,stroke-width:1px,stroke-dasharray:5,5
+    class USC_DESC,CTS_DESC,REQ_DESC,ETS_DESC,CH_DESC descBox
+```
+
+### Multiprocessing Communication Flow
+
+This diagram shows how processes communicate during a `run_model` execution:
+
+```mermaid
+sequenceDiagram
+    participant MP as Main Process<br/>(ClickHouse Server)
```

Comment:
technically isn't there also a queue between the main process and the forked child? I realize its more of an implementation detail rather htan an architectural point, but the layers of queues gets sort of confusing to reason about.

Reply:
added